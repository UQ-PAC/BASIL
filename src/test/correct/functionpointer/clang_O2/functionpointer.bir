000006f3: program
000006bb: sub __cxa_finalize(__cxa_finalize_result)
000006f4: __cxa_finalize_result :: out u32 = low:32[R0]

00000447:
0000053f: R16 := 0x11000
00000546: R17 := mem[R16 + 8, el]:u64
0000054c: R16 := R16 + 8
00000551: call R17 with noreturn

000006bc: sub __do_global_dtors_aux(__do_global_dtors_aux_result)
000006f5: __do_global_dtors_aux_result :: out u32 = low:32[R0]

000002be:
000002c2: #3 := R31 - 0x20
000002c8: mem := mem with [#3, el]:u64 <- R29
000002ce: mem := mem with [#3 + 8, el]:u64 <- R30
000002d2: R31 := #3
000002d8: R29 := R31
000002e0: mem := mem with [R31 + 0x10, el]:u64 <- R19
000002e5: R19 := 0x11000
000002ec: R0 := pad:64[mem[R19 + 0x30]]
000002f3: when 31:0[R0] <> 0 goto %000002f1
000006e9: goto %00000410

00000410:
00000413: R0 := 0x10000
0000041a: R0 := mem[R0 + 0xFC8, el]:u64
00000420: when R0 = 0 goto %0000041e
000006ea: goto %00000437

00000437:
0000043a: R0 := 0x11000
00000441: R0 := mem[R0 + 0x28, el]:u64
00000446: R30 := 0x6F0
00000449: call @__cxa_finalize with return %0000041e

0000041e:
00000426: R30 := 0x6F4
00000428: call @deregister_tm_clones with return %0000042a

0000042a:
0000042d: R0 := 1
00000435: mem := mem with [R19 + 0x30] <- 7:0[R0]
000006eb: goto %000002f1

000002f1:
000002fb: R19 := mem[R31 + 0x10, el]:u64
00000302: R29 := mem[R31, el]:u64
00000307: R30 := mem[R31 + 8, el]:u64
0000030b: R31 := R31 + 0x20
00000310: call R30 with noreturn

000006c0: sub __libc_start_main(__libc_start_main_main, __libc_start_main_arg2, __libc_start_main_arg3, __libc_start_main_auxv, __libc_start_main_result)
000006f6: __libc_start_main_main :: in u64 = R0
000006f7: __libc_start_main_arg2 :: in u32 = low:32[R1]
000006f8: __libc_start_main_arg3 :: in out u64 = R2
000006f9: __libc_start_main_auxv :: in out u64 = R3
000006fa: __libc_start_main_result :: out u32 = low:32[R0]

00000217:
00000529: R16 := 0x11000
00000530: R17 := mem[R16, el]:u64
00000536: R16 := R16
0000053b: call R17 with noreturn

000006c1: sub _fini(_fini_result)
000006fb: _fini_result :: out u32 = low:32[R0]

0000001f:
00000025: #0 := R31 - 0x10
0000002b: mem := mem with [#0, el]:u64 <- R29
00000031: mem := mem with [#0 + 8, el]:u64 <- R30
00000035: R31 := #0
0000003b: R29 := R31
00000042: R29 := mem[R31, el]:u64
00000047: R30 := mem[R31 + 8, el]:u64
0000004b: R31 := R31 + 0x10
00000050: call R30 with noreturn

000006c2: sub _init(_init_result)
000006fc: _init_result :: out u32 = low:32[R0]

00000601:
00000607: #8 := R31 - 0x10
0000060d: mem := mem with [#8, el]:u64 <- R29
00000613: mem := mem with [#8 + 8, el]:u64 <- R30
00000617: R31 := #8
0000061d: R29 := R31
00000622: R30 := 0x590
00000624: call @call_weak_fn with return %00000626

00000626:
0000062b: R29 := mem[R31, el]:u64
00000630: R30 := mem[R31 + 8, el]:u64
00000634: R31 := R31 + 0x10
00000639: call R30 with noreturn

000006c3: sub _start(_start_result)
000006fd: _start_result :: out u32 = low:32[R0]

000001d8:
000001dd: R29 := 0
000001e2: R30 := 0
000001e8: R5 := R0
000001ef: R1 := mem[R31, el]:u64
000001f5: R2 := R31 + 8
000001fb: R6 := R31
00000200: R0 := 0x10000
00000207: R0 := mem[R0 + 0xFD8, el]:u64
0000020c: R3 := 0
00000211: R4 := 0
00000216: R30 := 0x630
00000219: call @__libc_start_main with return %0000021b

0000021b:
0000021e: R30 := 0x634
00000221: call @abort with return %000006ec

000006ec:
000006ed: call @call_weak_fn with noreturn

000006c6: sub abort()


0000021f:
0000056b: R16 := 0x11000
00000572: R17 := mem[R16 + 0x18, el]:u64
00000578: R16 := R16 + 0x18
0000057d: call R17 with noreturn

000006c7: sub call_weak_fn(call_weak_fn_result)
000006fe: call_weak_fn_result :: out u32 = low:32[R0]

00000223:
00000226: R0 := 0x10000
0000022d: R0 := mem[R0 + 0xFD0, el]:u64
00000233: when R0 = 0 goto %00000231
000006ee: goto %00000487

00000231:
00000239: call R30 with noreturn

00000487:
0000048a: goto @__gmon_start__

00000488:
00000555: R16 := 0x11000
0000055c: R17 := mem[R16 + 0x10, el]:u64
00000562: R16 := R16 + 0x10
00000567: call R17 with noreturn

000006c9: sub deregister_tm_clones(deregister_tm_clones_result)
000006ff: deregister_tm_clones_result :: out u32 = low:32[R0]

0000023f:
00000242: R0 := 0x11000
00000248: R0 := R0 + 0x30
0000024d: R1 := 0x11000
00000253: R1 := R1 + 0x30
00000259: #1 := ~R0
0000025e: #2 := R1 + ~R0
00000264: VF := extend:65[#2 + 1] <> extend:65[R1] + extend:65[#1] + 1
0000026a: CF := pad:65[#2 + 1] <> pad:65[R1] + pad:65[#1] + 1
0000026e: ZF := #2 + 1 = 0
00000272: NF := 63:63[#2 + 1]
00000278: when ZF goto %00000276
000006ef: goto %00000469

00000469:
0000046c: R1 := 0x10000
00000473: R1 := mem[R1 + 0xFC0, el]:u64
00000478: when R1 = 0 goto %00000276
000006f0: goto %0000047c

00000276:
0000027e: call R30 with noreturn

0000047c:
00000480: R16 := R1
00000485: call R16 with noreturn

000006cc: sub frame_dummy(frame_dummy_result)
00000700: frame_dummy_result :: out u32 = low:32[R0]

00000316:
00000318: call @register_tm_clones with noreturn

000006cd: sub main(main_argc, main_argv, main_result)
00000701: main_argc :: in u32 = low:32[R0]
00000702: main_argv :: in out u64 = R1
00000703: main_result :: out u32 = low:32[R0]

0000035f:
00000363: #4 := R31 - 0x10
00000369: mem := mem with [#4, el]:u64 <- R29
0000036f: mem := mem with [#4 + 8, el]:u64 <- R30
00000373: R31 := #4
00000379: R29 := R31
0000037e: R8 := 0
00000384: R8 := R8 + 0x714
00000389: R9 := 0
0000038f: R9 := R9 + 0x724
00000395: #5 := 31:0[R0] - 2
0000039a: VF := extend:33[#5 + 1] <> extend:33[31:0[R0]] - 1
0000039f: CF := pad:33[#5 + 1] <> pad:33[31:0[R0]] + 0xFFFFFFFF
000003a3: ZF := #5 + 1 = 0
000003a7: NF := 31:31[#5 + 1]
000003ac: R10 := 0
000003b2: R10 := R10 + 0x734
000003c0: when ZF goto %000003b8
000003c1: goto %000003bc

000003bc:
000003bd: R8 := R8
000003c3: goto %000003bf

000003b8:
000003b9: R8 := R9
000003c2: goto %000003bf

000003bf:
000003c9: #6 := 31:0[R0] - 3
000003ce: VF := extend:33[#6 + 1] <> extend:33[31:0[R0]] - 2
000003d3: CF := pad:33[#6 + 1] <> pad:33[31:0[R0]] + 0xFFFFFFFE
000003d7: ZF := #6 + 1 = 0
000003db: NF := 31:31[#6 + 1]
000003e9: when ZF goto %000003e1
000003ea: goto %000003e5

000003e5:
000003e6: R8 := R8
000003ec: goto %000003e8

000003e1:
000003e2: R8 := R10
000003eb: goto %000003e8

000003e8:
000003f1: R30 := 0x778
000003f4: call R8 with return %000003f6

000003f6:
000003f9: R0 := 0
00000400: R29 := mem[R31, el]:u64
00000405: R30 := mem[R31 + 8, el]:u64
00000409: R31 := R31 + 0x10
0000040e: call R30 with noreturn

000006ce: sub register_tm_clones(register_tm_clones_result)
00000704: register_tm_clones_result :: out u32 = low:32[R0]

00000280:
00000283: R0 := 0x11000
00000289: R0 := R0 + 0x30
0000028e: R1 := 0x11000
00000294: R1 := R1 + 0x30
0000029b: R1 := R1 + ~R0 + 1
000002a1: R2 := 0.63:63[R1]
000002a8: R1 := R2 + (R1 ~>> 3)
000002ae: R1 := extend:64[63:1[R1]]
000002b4: when R1 = 0 goto %000002b2
000006f1: goto %0000044b

0000044b:
0000044e: R2 := 0x10000
00000455: R2 := mem[R2 + 0xFE0, el]:u64
0000045a: when R2 = 0 goto %000002b2
000006f2: goto %0000045e

000002b2:
000002ba: call R30 with noreturn

0000045e:
00000462: R16 := R2
00000467: call R16 with noreturn

000006d1: sub set_seven(set_seven_result)
00000705: set_seven_result :: out u32 = low:32[R0]

00000348:
0000034b: R8 := 0x11000
00000350: R9 := 7
00000358: mem := mem with [R8 + 0x34, el]:u32 <- 31:0[R9]
0000035d: call R30 with noreturn

000006d2: sub set_six(set_six_result)
00000706: set_six_result :: out u32 = low:32[R0]

00000331:
00000334: R8 := 0x11000
00000339: R9 := 6
00000341: mem := mem with [R8 + 0x34, el]:u32 <- 31:0[R9]
00000346: call R30 with noreturn

000006d3: sub set_two(set_two_result)
00000707: set_two_result :: out u32 = low:32[R0]

0000031a:
0000031d: R8 := 0x11000
00000322: R9 := 2
0000032a: mem := mem with [R8 + 0x34, el]:u32 <- 31:0[R9]
0000032f: call R30 with noreturn
