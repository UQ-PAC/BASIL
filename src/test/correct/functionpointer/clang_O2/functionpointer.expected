var {:extern} Gamma_R0: bool;
var {:extern} Gamma_R10: bool;
var {:extern} Gamma_R29: bool;
var {:extern} Gamma_R30: bool;
var {:extern} Gamma_R31: bool;
var {:extern} Gamma_R8: bool;
var {:extern} Gamma_R9: bool;
var {:extern} Gamma_mem: [bv64]bool;
var {:extern} Gamma_stack: [bv64]bool;
var {:extern} R0: bv64;
var {:extern} R10: bv64;
var {:extern} R29: bv64;
var {:extern} R30: bv64;
var {:extern} R31: bv64;
var {:extern} R8: bv64;
var {:extern} R9: bv64;
var {:extern} mem: [bv64]bv8;
var {:extern} stack: [bv64]bv8;
const {:extern} $_IO_stdin_used_addr: bv64;
axiom ($_IO_stdin_used_addr == 2008bv64);
const {:extern} $x_addr: bv64;
axiom ($x_addr == 131124bv64);
function {:extern} {:bvbuiltin "bvadd"} bvadd32(bv32, bv32) returns (bv32);
function {:extern} {:bvbuiltin "bvadd"} bvadd33(bv33, bv33) returns (bv33);
function {:extern} {:bvbuiltin "bvadd"} bvadd64(bv64, bv64) returns (bv64);
function {:extern} {:bvbuiltin "bvcomp"} bvcomp1(bv1, bv1) returns (bv1);
function {:extern} {:bvbuiltin "bvcomp"} bvcomp32(bv32, bv32) returns (bv1);
function {:extern} {:bvbuiltin "bvcomp"} bvcomp33(bv33, bv33) returns (bv1);
function {:extern} {:bvbuiltin "bvnot"} bvnot1(bv1) returns (bv1);
function {:extern} gamma_load64(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 7bv64)] && (gammaMap[bvadd64(index, 6bv64)] && (gammaMap[bvadd64(index, 5bv64)] && (gammaMap[bvadd64(index, 4bv64)] && (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))))))
}

function {:extern} gamma_store64(gammaMap: [bv64]bool, index: bv64, value: bool) returns ([bv64]bool) {
  gammaMap[index := value][bvadd64(index, 1bv64) := value][bvadd64(index, 2bv64) := value][bvadd64(index, 3bv64) := value][bvadd64(index, 4bv64) := value][bvadd64(index, 5bv64) := value][bvadd64(index, 6bv64) := value][bvadd64(index, 7bv64) := value]
}

function {:extern} memory_load64_le(memory: [bv64]bv8, index: bv64) returns (bv64) {
  (memory[bvadd64(index, 7bv64)] ++ (memory[bvadd64(index, 6bv64)] ++ (memory[bvadd64(index, 5bv64)] ++ (memory[bvadd64(index, 4bv64)] ++ (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))))))
}

function {:extern} memory_load8_le(memory: [bv64]bv8, index: bv64) returns (bv8) {
  memory[index]
}

function {:extern} memory_store64_le(memory: [bv64]bv8, index: bv64, value: bv64) returns ([bv64]bv8) {
  memory[index := value[8:0]][bvadd64(index, 1bv64) := value[16:8]][bvadd64(index, 2bv64) := value[24:16]][bvadd64(index, 3bv64) := value[32:24]][bvadd64(index, 4bv64) := value[40:32]][bvadd64(index, 5bv64) := value[48:40]][bvadd64(index, 6bv64) := value[56:48]][bvadd64(index, 7bv64) := value[64:56]]
}

function {:extern} {:bvbuiltin "sign_extend 1"} sign_extend1_32(bv32) returns (bv33);
function {:extern} {:bvbuiltin "zero_extend 1"} zero_extend1_32(bv32) returns (bv33);
procedure {:extern} rely();
  modifies Gamma_mem, mem;
  ensures (mem == old(mem));
  ensures (Gamma_mem == old(Gamma_mem));
  free ensures (memory_load8_le(mem, 2008bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 2009bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 2010bv64) == 2bv8);
  free ensures (memory_load8_le(mem, 2011bv64) == 0bv8);
  free ensures (memory_load64_le(mem, 130488bv64) == 1872bv64);
  free ensures (memory_load64_le(mem, 130496bv64) == 1792bv64);
  free ensures (memory_load64_le(mem, 131032bv64) == 1924bv64);
  free ensures (memory_load64_le(mem, 131112bv64) == 131112bv64);

procedure {:extern} rely_transitive();
  modifies Gamma_mem, mem;
  ensures (mem == old(mem));
  ensures (Gamma_mem == old(Gamma_mem));

implementation {:extern} rely_transitive()
{
  call rely();
  call rely();
}

procedure {:extern} rely_reflexive();

procedure {:extern} guarantee_reflexive();
  modifies Gamma_mem, mem;

procedure main();
  modifies Gamma_R0, Gamma_R10, Gamma_R29, Gamma_R30, Gamma_R31, Gamma_R8, Gamma_R9, Gamma_stack, R0, R10, R29, R30, R31, R8, R9, stack;
  free requires (memory_load64_le(mem, 131104bv64) == 0bv64);
  free requires (memory_load64_le(mem, 131112bv64) == 131112bv64);
  free requires (memory_load8_le(mem, 2008bv64) == 1bv8);
  free requires (memory_load8_le(mem, 2009bv64) == 0bv8);
  free requires (memory_load8_le(mem, 2010bv64) == 2bv8);
  free requires (memory_load8_le(mem, 2011bv64) == 0bv8);
  free requires (memory_load64_le(mem, 130488bv64) == 1872bv64);
  free requires (memory_load64_le(mem, 130496bv64) == 1792bv64);
  free requires (memory_load64_le(mem, 131032bv64) == 1924bv64);
  free requires (memory_load64_le(mem, 131112bv64) == 131112bv64);
  free ensures (Gamma_R29 == old(Gamma_R29));
  free ensures (Gamma_R31 == old(Gamma_R31));
  free ensures (R29 == old(R29));
  free ensures (R31 == old(R31));
  free ensures (memory_load8_le(mem, 2008bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 2009bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 2010bv64) == 2bv8);
  free ensures (memory_load8_le(mem, 2011bv64) == 0bv8);
  free ensures (memory_load64_le(mem, 130488bv64) == 1872bv64);
  free ensures (memory_load64_le(mem, 130496bv64) == 1792bv64);
  free ensures (memory_load64_le(mem, 131032bv64) == 1924bv64);
  free ensures (memory_load64_le(mem, 131112bv64) == 131112bv64);

implementation main()
{
  var #4: bv64;
  var #5: bv32;
  var #6: bv32;
  var CF: bv1;
  var Gamma_#4: bool;
  var Gamma_#5: bool;
  var Gamma_#6: bool;
  var Gamma_CF: bool;
  var Gamma_NF: bool;
  var Gamma_VF: bool;
  var Gamma_ZF: bool;
  var NF: bv1;
  var VF: bv1;
  var ZF: bv1;
  lmain:
    assume {:captureState "lmain"} true;
    #4, Gamma_#4 := bvadd64(R31, 18446744073709551600bv64), Gamma_R31;
    stack, Gamma_stack := memory_store64_le(stack, #4, R29), gamma_store64(Gamma_stack, #4, Gamma_R29);
    assume {:captureState "%00000369"} true;
    stack, Gamma_stack := memory_store64_le(stack, bvadd64(#4, 8bv64), R30), gamma_store64(Gamma_stack, bvadd64(#4, 8bv64), Gamma_R30);
    assume {:captureState "%0000036f"} true;
    R31, Gamma_R31 := #4, Gamma_#4;
    R29, Gamma_R29 := R31, Gamma_R31;
    R8, Gamma_R8 := 0bv64, true;
    R8, Gamma_R8 := bvadd64(R8, 1876bv64), Gamma_R8;
    R9, Gamma_R9 := 0bv64, true;
    R9, Gamma_R9 := bvadd64(R9, 1892bv64), Gamma_R9;
    #5, Gamma_#5 := bvadd32(R0[32:0], 4294967294bv32), Gamma_R0;
    VF, Gamma_VF := bvnot1(bvcomp33(sign_extend1_32(bvadd32(#5, 1bv32)), bvadd33(sign_extend1_32(R0[32:0]), 8589934591bv33))), (Gamma_R0 && Gamma_#5);
    CF, Gamma_CF := bvnot1(bvcomp33(zero_extend1_32(bvadd32(#5, 1bv32)), bvadd33(zero_extend1_32(R0[32:0]), 4294967295bv33))), (Gamma_R0 && Gamma_#5);
    ZF, Gamma_ZF := bvcomp32(bvadd32(#5, 1bv32), 0bv32), Gamma_#5;
    NF, Gamma_NF := bvadd32(#5, 1bv32)[32:31], Gamma_#5;
    R10, Gamma_R10 := 0bv64, true;
    R10, Gamma_R10 := bvadd64(R10, 1908bv64), Gamma_R10;
    assert Gamma_ZF;
    goto lmain_goto_l000003b8, lmain_goto_l000003bc;
  l000003bc:
    assume {:captureState "l000003bc"} true;
    R8, Gamma_R8 := R8, Gamma_R8;
    goto l000003bf;
  l000003b8:
    assume {:captureState "l000003b8"} true;
    R8, Gamma_R8 := R9, Gamma_R9;
    goto l000003bf;
  l000003bf:
    assume {:captureState "l000003bf"} true;
    #6, Gamma_#6 := bvadd32(R0[32:0], 4294967293bv32), Gamma_R0;
    VF, Gamma_VF := bvnot1(bvcomp33(sign_extend1_32(bvadd32(#6, 1bv32)), bvadd33(sign_extend1_32(R0[32:0]), 8589934590bv33))), (Gamma_R0 && Gamma_#6);
    CF, Gamma_CF := bvnot1(bvcomp33(zero_extend1_32(bvadd32(#6, 1bv32)), bvadd33(zero_extend1_32(R0[32:0]), 4294967294bv33))), (Gamma_R0 && Gamma_#6);
    ZF, Gamma_ZF := bvcomp32(bvadd32(#6, 1bv32), 0bv32), Gamma_#6;
    NF, Gamma_NF := bvadd32(#6, 1bv32)[32:31], Gamma_#6;
    assert Gamma_ZF;
    goto l000003bf_goto_l000003e5, l000003bf_goto_l000003e1;
  l000003e5:
    assume {:captureState "l000003e5"} true;
    R8, Gamma_R8 := R8, Gamma_R8;
    goto l000003e8;
  l000003e1:
    assume {:captureState "l000003e1"} true;
    R8, Gamma_R8 := R10, Gamma_R10;
    goto l000003e8;
  l000003e8:
    assume {:captureState "l000003e8"} true;
    R30, Gamma_R30 := 1976bv64, true;
    //UNRESOLVED: call R8
    assert false;
    goto l000003f6;
  l000003f6:
    assume {:captureState "l000003f6"} true;
    R0, Gamma_R0 := 0bv64, true;
    R29, Gamma_R29 := memory_load64_le(stack, R31), gamma_load64(Gamma_stack, R31);
    R30, Gamma_R30 := memory_load64_le(stack, bvadd64(R31, 8bv64)), gamma_load64(Gamma_stack, bvadd64(R31, 8bv64));
    R31, Gamma_R31 := bvadd64(R31, 16bv64), Gamma_R31;
    goto main_return;
  lmain_goto_l000003b8:
    assume {:captureState "lmain_goto_l000003b8"} true;
    assume (bvcomp1(ZF, 1bv1) != 0bv1);
    goto l000003b8;
  lmain_goto_l000003bc:
    assume {:captureState "lmain_goto_l000003bc"} true;
    assume (bvcomp1(ZF, 1bv1) == 0bv1);
    goto l000003bc;
  l000003bf_goto_l000003e1:
    assume {:captureState "l000003bf_goto_l000003e1"} true;
    assume (bvcomp1(ZF, 1bv1) != 0bv1);
    goto l000003e1;
  l000003bf_goto_l000003e5:
    assume {:captureState "l000003bf_goto_l000003e5"} true;
    assume (bvcomp1(ZF, 1bv1) == 0bv1);
    goto l000003e5;
  main_return:
    assume {:captureState "main_return"} true;
    return;
}

